src/schema/schema_alter.c

```diff
@@ -63,7 +54,166 @@ err:	__wt_free(session, config);
 	if (ret == WT_NOTFOUND)
 		ret = ENOENT;
 
-	WT_TRET(__wt_meta_track_off(session, true, ret != 0));
+	return (ret);
+}
+
+/*
+ * __alter_file --
+ *	Alter a file.
+ */
+static int
+__alter_file(WT_SESSION_IMPL *session, const char *newcfg[])
+{
+	WT_DECL_RET;
+	const char *filename, *uri;
+
+	/*
+	 * We know that we have exclusive access to the file.  So it will be
+	 * closed after we're done with it and the next open will see the
+	 * updated metadata.
+	 */
+	filename = uri = session->dhandle->name;
+	if (!WT_PREFIX_SKIP(filename, "file:"))
+		return (__wt_unexpected_object_type(session, uri, "file:"));
+
+	WT_RET(__wt_meta_track_on(session));
```

Comment:
I would expect the calls to `__wt_meta_track_on/off` to be in the top-level function (`__wt_schema_alter`) rather than just for `file:` types.

In case it isn't clear, this is like "begin transaction" for schema-changing operations: we want the whole operation wrapped in a transaction so that if it fails part-way through, we roll back all of the changes.

Reply:
I did look at putting meta tracking around all the changes. What happens is once meta tracking is switched ON, when file alter tries to lock the dhandle in fails.
 (`__wt_session_get_dhandle` called by `__wt_session_get_btree_ckpt` called by `__wt_exclusive_handle_operation`).

I will work with @agorrod to figure out a solution to this problem.